File: problems/java/math_geometry/rotate_image.java
class Solution 
{
    public void rotate(int[][] matrix) 
    {
        int size        = matrix.length;
        int [][] tmp    = new int [size][size];

        for(int i=0; i<size; i++)
            for(int j=0; j<size; j++)
                tmp[i][j] = matrix[i][j];
        
        for(int y=0; y<size; y++)
            for(int x=0; x<size; x++)
                matrix[y][x] = tmp[size-x-1][y];
    }
}

File: python/geometry/rotate_image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        l, r = 0, len(matrix)-1

        while l < r:
            for i in range(r-l):
                start, end = l, r

                temp = matrix[start][start+i]

                matrix[start][start+i] = matrix[end-i][start]
                matrix[end-i][start] = matrix[end][end-i]
                matrix[end][end-i] = matrix[start+i][end]
                matrix[start+i][end] = temp
            l += 1
            r -= 1